Cannot connect from Wyse Thin clients

Hey All,

I have stumbled across this issue https://github.com/neutrinolabs/xrdp/issues/874 while researching an issue I have been having with some Wyse sx0 thin clients and I believe it is the same issue.

The original issue was closed because the developers were unable to get a hold of one of the clients, I have a box of them and would be happy to ship a few out if it could help fix this issue.

Alright, had some free time this evening and compiled the latest version from the repo:

```
xrdp --version
Argument 0 - xrdp
Argument 1 - --version
xrdp 0.9.18
  A Remote Desktop Protocol Server.
  Copyright (C) 2004-2020 Jay Sorg, Neutrino Labs, and all contributors.
  See https://github.com/neutrinolabs/xrdp for more information.

  Configure options:
      --enable-devel-all

  Compiled with OpenSSL 1.1.1f  31 Mar 2020

```

Tried connecting from the thin client and getting the following log:
```
[20220304-21:28:41] [INFO ] [g_sck_accept(os_calls.c:1384)] Socket 12: AF_INET connection received from 10.13.37.205 port 3244
[20220304-21:28:41] [INFO ] [xrdp_rdp_read_config(xrdp_rdp.c:225)] Using default X.509 certificate: /etc/xrdp/cert.pem
[20220304-21:28:41] [INFO ] [xrdp_rdp_read_config(xrdp_rdp.c:257)] Using default X.509 key file: /etc/xrdp/key.pem
[20220304-21:28:41] [INFO ] [xrdp_iso_negotiate_security(xrdp_iso.c:194)] Security protocol: configured [SSL|RDP], requested [RDP], selected [RDP]
[20220304-21:28:41] [INFO ] [xrdp_sec_process_mcs_data_CS_CORE(xrdp_sec.c:1974)] Connected client computer name: WT008064857377
[20220304-21:28:41] [ERROR] [xrdp_mcs_recv_cjrq(xrdp_mcs.c:746)] Parsed [ITU-T T.125] DomainMCSPDU choice index expected 14, received 25
[20220304-21:28:41] [ERROR] [xrdp_mcs_incoming(xrdp_mcs.c:1242)] [MCS Connection Sequence] receive channel join request failed
[20220304-21:28:41] [ERROR] [xrdp_sec_incoming(xrdp_sec.c:2869)] xrdp_sec_incoming: xrdp_mcs_incoming failed
[20220304-21:28:41] [ERROR] [xrdp_rdp_incoming(xrdp_rdp.c:921)] xrdp_rdp_incoming: xrdp_sec_incoming failed
[20220304-21:28:41] [ERROR] [xrdp_process_main_loop(xrdp_process.c:302)] xrdp_process_main_loop: libxrdp_process_incoming failed
```

Let me know what else I can provide

I just realized when I ran make install the crypt and security_layer were reset.  I have put them back to RDP and low, here's the latest log

```
[20220304-21:31:23] [INFO ] [g_sck_accept(os_calls.c:1384)] Socket 12: AF_INET connection received from 10.13.37.205 port 3247
[20220304-21:31:23] [INFO ] [xrdp_rdp_read_config(xrdp_rdp.c:225)] Using default X.509 certificate: /etc/xrdp/cert.pem
[20220304-21:31:23] [INFO ] [xrdp_rdp_read_config(xrdp_rdp.c:257)] Using default X.509 key file: /etc/xrdp/key.pem
[20220304-21:31:23] [INFO ] [xrdp_iso_negotiate_security(xrdp_iso.c:194)] Security protocol: configured [RDP], requested [RDP], selected [RDP]
[20220304-21:31:23] [INFO ] [xrdp_sec_process_mcs_data_CS_CORE(xrdp_sec.c:1974)] Connected client computer name: WT008064857377
[20220304-21:31:23] [ERROR] [xrdp_mcs_recv_cjrq(xrdp_mcs.c:746)] Parsed [ITU-T T.125] DomainMCSPDU choice index expected 14, received 25
[20220304-21:31:23] [ERROR] [xrdp_mcs_incoming(xrdp_mcs.c:1242)] [MCS Connection Sequence] receive channel join request failed
[20220304-21:31:23] [ERROR] [xrdp_sec_incoming(xrdp_sec.c:2869)] xrdp_sec_incoming: xrdp_mcs_incoming failed
[20220304-21:31:23] [ERROR] [xrdp_rdp_incoming(xrdp_rdp.c:921)] xrdp_rdp_incoming: xrdp_sec_incoming failed
[20220304-21:31:23] [ERROR] [xrdp_process_main_loop(xrdp_process.c:302)] xrdp_process_main_loop: libxrdp_process_incoming failed
```

@TOMATO-ONE - that's really helpful. Thanks for your input.

@christheradioguy - the errors you're getting are not related to encryption, but rather to the connection setup sequence to do with channels.

Referring to section [1.3.1.1 of [MS-RDPBCGR]](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpbcgr/023f1e69-cfe8-4ee6-9ee0-7e759fb4e4ee), xrdp sends an 'MCS Attach User Confirm PDU`, and then expects a number of 
'Channel Join Request PDUs'. This message is telling us that we've got a 'send data request PDU' where we were expecting a 'Channel join request PDU':-

```
[20220304-21:31:23] [ERROR] [xrdp_mcs_recv_cjrq(xrdp_mcs.c:746)] Parsed [ITU-T T.125] DomainMCSPDU choice index expected 14, received 25
```

Looking at the code, it looks like the thin client thinks there are less channels than the server is expecting to support.

We can get more logging if you set the logging level to DEBUG. Or you could try playing around with the requested facilities on the ThinOS (although I have no idea how you'd do that).

I can't find any on-line release notes for ThinOS that old, so I'm unable to advise you on what extra facilities may be present on the newer firmware.

Alright, here's what I get with DEBUG mode:

```
[20220404-19:26:37] [INFO ] setting send buffer to 262144 bytes
[20220404-19:26:37] [INFO ] send buffer set to 425984 bytes
[20220404-19:26:37] [INFO ] xrdp_listen_pp done
[20220404-19:26:37] [DEBUG] Closed socket 7 (AF_INET 0.0.0.0:3389)
[20220404-19:26:39] [INFO ] starting xrdp with pid 34088
[20220404-19:26:39] [INFO ] address [0.0.0.0] port [3389] mode 1
[20220404-19:26:39] [INFO ] listening to port 3389 on 0.0.0.0
[20220404-19:26:39] [INFO ] setting send buffer to 262144 bytes
[20220404-19:26:39] [INFO ] send buffer set to 425984 bytes
[20220404-19:26:39] [INFO ] xrdp_listen_pp done
[20220404-19:35:15] [INFO ] Socket 12: AF_INET connection received from 10.13.37.123 port 3400
[20220404-19:35:15] [DEBUG] Closed socket 12 (AF_INET 10.4.9.34:3389)
[20220404-19:35:15] [DEBUG] Closed socket 11 (AF_INET 0.0.0.0:3389)
[20220404-19:35:15] [DEBUG] item ini_version, value 1
[20220404-19:35:15] [DEBUG] item fork, value true
[20220404-19:35:15] [DEBUG] item port, value 3389
[20220404-19:35:15] [DEBUG] item use_vsock, value false
[20220404-19:35:15] [DEBUG] item tcp_nodelay, value false
[20220404-19:35:15] [DEBUG] item tcp_keepalive, value true
[20220404-19:35:15] [DEBUG] item tcp_send_buffer_bytes, value 262144
[20220404-19:35:15] [DEBUG] item security_layer, value negotiate
[20220404-19:35:15] [DEBUG] item crypt_level, value low
[20220404-19:35:15] [DEBUG] item certificate, value 
[20220404-19:35:15] [INFO ] Using default X.509 certificate: /etc/xrdp/cert.pem
[20220404-19:35:15] [DEBUG] item key_file, value 
[20220404-19:35:15] [INFO ] Using default X.509 key file: /etc/xrdp/key.pem
[20220404-19:35:15] [DEBUG] item ssl_protocols, value TLSv1.2, TLSv1.3
[20220404-19:35:15] [DEBUG] TLSv1.3 enabled
[20220404-19:35:15] [DEBUG] TLSv1.2 enabled
[20220404-19:35:15] [DEBUG] item autorun, value 
[20220404-19:35:15] [DEBUG] item allow_channels, value true
[20220404-19:35:15] [DEBUG] item allow_multimon, value true
[20220404-19:35:15] [DEBUG] item bitmap_cache, value true
[20220404-19:35:15] [DEBUG] item bitmap_compression, value true
[20220404-19:35:15] [DEBUG] item bulk_compression, value true
[20220404-19:35:15] [DEBUG] item max_bpp, value 128
[20220404-19:35:15] [DEBUG] item xserverbpp, value 128
[20220404-19:35:15] [DEBUG] item use_compression, value yes
[20220404-19:35:15] [DEBUG] item new_cursors, value true
[20220404-19:35:15] [DEBUG] item use_fastpath, value both
[20220404-19:35:15] [DEBUG] item blue, value 009cb5
[20220404-19:35:15] [DEBUG] item grey, value dedede
[20220404-19:35:15] [DEBUG] item ls_top_window_bg_color, value 009cb5
[20220404-19:35:15] [DEBUG] item ls_width, value 350
[20220404-19:35:15] [DEBUG] item ls_height, value 430
[20220404-19:35:15] [DEBUG] item ls_bg_color, value dedede
[20220404-19:35:15] [DEBUG] item ls_logo_filename, value 
[20220404-19:35:15] [DEBUG] item ls_logo_x_pos, value 55
[20220404-19:35:15] [DEBUG] item ls_logo_y_pos, value 50
[20220404-19:35:15] [DEBUG] item ls_label_x_pos, value 30
[20220404-19:35:15] [DEBUG] item ls_label_width, value 65
[20220404-19:35:15] [DEBUG] item ls_input_x_pos, value 110
[20220404-19:35:15] [DEBUG] item ls_input_width, value 210
[20220404-19:35:15] [DEBUG] item ls_input_y_pos, value 220
[20220404-19:35:15] [DEBUG] item ls_btn_ok_x_pos, value 142
[20220404-19:35:15] [DEBUG] item ls_btn_ok_y_pos, value 370
[20220404-19:35:15] [DEBUG] item ls_btn_ok_width, value 85
[20220404-19:35:15] [DEBUG] item ls_btn_ok_height, value 30
[20220404-19:35:15] [DEBUG] item ls_btn_cancel_x_pos, value 237
[20220404-19:35:15] [DEBUG] item ls_btn_cancel_y_pos, value 370
[20220404-19:35:15] [DEBUG] item ls_btn_cancel_width, value 85
[20220404-19:35:15] [DEBUG] item ls_btn_cancel_height, value 30
[20220404-19:35:15] [INFO ] Security protocol: configured [SSL|RDP], requested [RDP], selected [RDP]
[20220404-19:35:15] [DEBUG] Using RDP security, and reading the server configuration
[20220404-19:35:15] [DEBUG] [MCS Connection Sequence] receive connection request
[20220404-19:35:15] [INFO ] Connected client computer name: WT00806465d8c3
[20220404-19:35:15] [DEBUG] Client supports 40 bit encryption
[20220404-19:35:15] [DEBUG] Client supports 128 bit encryption
[20220404-19:35:15] [DEBUG] Client supports 56 bit encryption
[20220404-19:35:15] [DEBUG] Client supports fips encryption
[20220404-19:35:15] [DEBUG] Client and server both support low encryption, using RDP 40-bit encryption.
[20220404-19:35:15] [DEBUG] Adding channel: name rdpdr, channel id 1004, flags 0x82000000
[20220404-19:35:15] [DEBUG] Adding channel: name cliprdr, channel id 1005, flags 0x98000000
[20220404-19:35:15] [DEBUG] Adding channel: name CUVA, channel id 1006, flags 0x88000000
[20220404-19:35:15] [DEBUG] Adding channel: name AURTCX, channel id 1007, flags 0x88000000
[20220404-19:35:15] [DEBUG] Adding channel: name USBVDX, channel id 1008, flags 0x88000000
[20220404-19:35:15] [DEBUG] Adding channel: name CARANOC, channel id 1009, flags 0x82000000
[20220404-19:35:15] [DEBUG] [MCS Connection Sequence] construct connection reponse
[20220404-19:35:15] [DEBUG] using 2048 bit RSA key
[20220404-19:35:15] [DEBUG] [MCS Connection Sequence] send connection reponse
[20220404-19:35:15] [DEBUG] [MCS Connection Sequence] receive erect domain request
[20220404-19:35:15] [DEBUG] [MCS Connection Sequence] receive attach user request
[20220404-19:35:15] [DEBUG] [MCS Connection Sequence] send attach user confirm
[20220404-19:35:15] [DEBUG] [MCS Connection Sequence] receive channel join request
[20220404-19:35:15] [DEBUG] [MCS Connection Sequence] send channel join confirm
[20220404-19:35:15] [DEBUG] [MCS Connection Sequence] receive channel join request
[20220404-19:35:15] [DEBUG] [MCS Connection Sequence] send channel join confirm
[20220404-19:35:15] [DEBUG] [MCS Connection Sequence] receive channel join request
[20220404-19:35:15] [DEBUG] [MCS Connection Sequence] send channel join confirm
[20220404-19:35:15] [DEBUG] [MCS Connection Sequence] receive channel join request
[20220404-19:35:15] [DEBUG] [MCS Connection Sequence] send channel join confirm
[20220404-19:35:15] [DEBUG] [MCS Connection Sequence] receive channel join request
[20220404-19:35:15] [DEBUG] [MCS Connection Sequence] send channel join confirm
[20220404-19:35:15] [DEBUG] [MCS Connection Sequence] receive channel join request
[20220404-19:35:15] [DEBUG] [MCS Connection Sequence] send channel join confirm
[20220404-19:35:15] [DEBUG] [MCS Connection Sequence] receive channel join request
[20220404-19:35:15] [DEBUG] [MCS Connection Sequence] send channel join confirm
[20220404-19:35:15] [DEBUG] [MCS Connection Sequence] receive channel join request
[20220404-19:35:16] [ERROR] Parsed [ITU-T T.125] DomainMCSPDU choice index expected 14, received 25
[20220404-19:35:16] [ERROR] [MCS Connection Sequence] receive channel join request failed
[20220404-19:35:16] [ERROR] xrdp_sec_incoming: xrdp_mcs_incoming failed
[20220404-19:35:16] [ERROR] xrdp_rdp_incoming: xrdp_sec_incoming failed
[20220404-19:35:16] [ERROR] xrdp_process_main_loop: libxrdp_process_incoming failed
[20220404-19:35:16] [DEBUG] Closed socket 12 (AF_INET 10.4.9.34:3389)
```

Apologies, I forgot to compile with the `--enable-devel-all` argument.  The output looks pretty simialr though.

This is the version I am running

```
# xrdp --version
Argument 0 - xrdp
Argument 1 - --version
xrdp 0.9.19
  A Remote Desktop Protocol Server.
  Copyright (C) 2004-2020 Jay Sorg, Neutrino Labs, and all contributors.
  See https://github.com/neutrinolabs/xrdp for more information.

  Configure options:
      --enable-devel-all

  Compiled with OpenSSL 1.1.1f  31 Mar 2020

```

```
[20220406-17:32:56] [INFO ] [xrdp_listen_process_startup_params(xrdp_listen.c:719)] setting send buffer to 262144 bytes
[20220406-17:32:56] [INFO ] [xrdp_listen_process_startup_params(xrdp_listen.c:734)] send buffer set to 425984 bytes
[20220406-17:32:56] [INFO ] [xrdp_listen_process_startup_params(xrdp_listen.c:674)] xrdp_listen_pp done
[20220406-17:32:56] [DEBUG] [g_sck_close(os_calls.c:766)] Closed socket 7 (AF_INET 0.0.0.0:3389)
[20220406-17:32:58] [INFO ] [main(xrdp.c:685)] starting xrdp with pid 54331
[20220406-17:32:58] [INFO ] [xrdp_listen_process_startup_params(xrdp_listen.c:678)] address [0.0.0.0] port [3389] mode 1
[20220406-17:32:58] [INFO ] [xrdp_listen_process_startup_params(xrdp_listen.c:687)] listening to port 3389 on 0.0.0.0
[20220406-17:32:58] [INFO ] [xrdp_listen_process_startup_params(xrdp_listen.c:719)] setting send buffer to 262144 bytes
[20220406-17:32:58] [INFO ] [xrdp_listen_process_startup_params(xrdp_listen.c:734)] send buffer set to 425984 bytes
[20220406-17:32:58] [INFO ] [xrdp_listen_process_startup_params(xrdp_listen.c:674)] xrdp_listen_pp done
[20220406-17:34:33] [INFO ] [g_sck_accept(os_calls.c:1395)] Socket 12: AF_INET connection received from 10.13.37.123 port 3477
[20220406-17:34:33] [DEBUG] [g_sck_close(os_calls.c:766)] Closed socket 12 (AF_INET 10.4.9.34:3389)
[20220406-17:34:33] [DEBUG] [g_sck_close(os_calls.c:766)] Closed socket 11 (AF_INET 0.0.0.0:3389)
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ini_version, value 1
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item fork, value true
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item port, value 3389
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item use_vsock, value false
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item tcp_nodelay, value false
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item tcp_keepalive, value true
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item tcp_send_buffer_bytes, value 262144
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item security_layer, value negotiate
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item crypt_level, value low
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item certificate, value 
[20220406-17:34:33] [INFO ] [xrdp_rdp_read_config(xrdp_rdp.c:225)] Using default X.509 certificate: /etc/xrdp/cert.pem
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item key_file, value 
[20220406-17:34:33] [INFO ] [xrdp_rdp_read_config(xrdp_rdp.c:257)] Using default X.509 key file: /etc/xrdp/key.pem
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ssl_protocols, value TLSv1.2, TLSv1.3
[20220406-17:34:33] [DEBUG] [ssl_get_protocols_from_string(ssl_calls.c:1458)] TLSv1.3 enabled
[20220406-17:34:33] [DEBUG] [ssl_get_protocols_from_string(ssl_calls.c:1470)] TLSv1.2 enabled
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item autorun, value 
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item allow_channels, value true
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item allow_multimon, value true
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item bitmap_cache, value true
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item bitmap_compression, value true
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item bulk_compression, value true
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item max_bpp, value 128
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item xserverbpp, value 128
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item use_compression, value yes
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item new_cursors, value true
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item use_fastpath, value both
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item blue, value 009cb5
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item grey, value dedede
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_top_window_bg_color, value 009cb5
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_width, value 350
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_height, value 430
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_bg_color, value dedede
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_logo_filename, value 
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_logo_x_pos, value 55
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_logo_y_pos, value 50
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_label_x_pos, value 30
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_label_width, value 65
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_input_x_pos, value 110
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_input_width, value 210
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_input_y_pos, value 220
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_btn_ok_x_pos, value 142
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_btn_ok_y_pos, value 370
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_btn_ok_width, value 85
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_btn_ok_height, value 30
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_btn_cancel_x_pos, value 237
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_btn_cancel_y_pos, value 370
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_btn_cancel_width, value 85
[20220406-17:34:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_btn_cancel_height, value 30
[20220406-17:34:33] [DEBUG] [xrdp_iso_incoming(xrdp_iso.c:485)] [ITU-T X.224] Connection Sequence: receive connection request
[20220406-17:34:33] [INFO ] [xrdp_iso_negotiate_security(xrdp_iso.c:194)] Security protocol: configured [SSL|RDP], requested [RDP], selected [RDP]
[20220406-17:34:33] [DEBUG] [xrdp_iso_incoming(xrdp_iso.c:587)] [ITU-T X.224] Connection Sequence: send connection confirmation
[20220406-17:34:33] [DEBUG] [xrdp_iso_incoming(xrdp_iso.c:594)] [ITU-T X.224] Connection Sequence: completed
[20220406-17:34:33] [DEBUG] [xrdp_sec_incoming(xrdp_sec.c:2709)] Using RDP security, and reading the server configuration
[20220406-17:34:33] [DEBUG] [xrdp_sec_incoming(xrdp_sec.c:2742)] Server config: pub_mod bytes 256
[20220406-17:34:33] [DEBUG] [xrdp_sec_incoming(xrdp_sec.c:2753)] Server config: pri_exp 256
[20220406-17:34:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1188)] [MCS Connection Sequence] receive connection request
[20220406-17:34:33] [INFO ] [xrdp_sec_process_mcs_data_CS_CORE(xrdp_sec.c:1972)] Connected client computer name: WT00806465d8c3
[20220406-17:34:33] [DEBUG] [xrdp_sec_process_mcs_data_CS_SECURITY(xrdp_sec.c:2151)] Client supports 40 bit encryption
[20220406-17:34:33] [DEBUG] [xrdp_sec_process_mcs_data_CS_SECURITY(xrdp_sec.c:2155)] Client supports 128 bit encryption
[20220406-17:34:33] [DEBUG] [xrdp_sec_process_mcs_data_CS_SECURITY(xrdp_sec.c:2159)] Client supports 56 bit encryption
[20220406-17:34:33] [DEBUG] [xrdp_sec_process_mcs_data_CS_SECURITY(xrdp_sec.c:2163)] Client supports fips encryption
[20220406-17:34:33] [DEBUG] [xrdp_sec_process_mcs_data_CS_SECURITY(xrdp_sec.c:2225)] Client and server both support low encryption, using RDP 40-bit encryption.
[20220406-17:34:33] [DEBUG] [xrdp_sec_process_mcs_data(xrdp_sec.c:2460)] Received [MS-RDPBCGR] TS_UD_CS_CLUSTER - no-op
[20220406-17:34:33] [DEBUG] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2310)] Adding channel: name rdpdr, channel id 1004, flags 0x82000000
[20220406-17:34:33] [DEBUG] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2310)] Adding channel: name cliprdr, channel id 1005, flags 0x98000000
[20220406-17:34:33] [DEBUG] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2310)] Adding channel: name CUVA, channel id 1006, flags 0x88000000
[20220406-17:34:33] [DEBUG] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2310)] Adding channel: name AURTCX, channel id 1007, flags 0x88000000
[20220406-17:34:33] [DEBUG] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2310)] Adding channel: name USBVDX, channel id 1008, flags 0x88000000
[20220406-17:34:33] [DEBUG] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2310)] Adding channel: name CARANOC, channel id 1009, flags 0x82000000
[20220406-17:34:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1202)] [MCS Connection Sequence] construct connection reponse
[20220406-17:34:33] [DEBUG] [xrdp_mcs_out_gcc_data(xrdp_mcs.c:1038)] using 2048 bit RSA key
[20220406-17:34:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1209)] [MCS Connection Sequence] send connection reponse
[20220406-17:34:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1216)] [MCS Connection Sequence] receive erect domain request
[20220406-17:34:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1223)] [MCS Connection Sequence] receive attach user request
[20220406-17:34:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1230)] [MCS Connection Sequence] send attach user confirm
[20220406-17:34:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1239)] [MCS Connection Sequence] receive channel join request
[20220406-17:34:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1246)] [MCS Connection Sequence] send channel join confirm
[20220406-17:34:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1239)] [MCS Connection Sequence] receive channel join request
[20220406-17:34:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1246)] [MCS Connection Sequence] send channel join confirm
[20220406-17:34:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1239)] [MCS Connection Sequence] receive channel join request
[20220406-17:34:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1246)] [MCS Connection Sequence] send channel join confirm
[20220406-17:34:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1239)] [MCS Connection Sequence] receive channel join request
[20220406-17:34:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1246)] [MCS Connection Sequence] send channel join confirm
[20220406-17:34:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1239)] [MCS Connection Sequence] receive channel join request
[20220406-17:34:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1246)] [MCS Connection Sequence] send channel join confirm
[20220406-17:34:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1239)] [MCS Connection Sequence] receive channel join request
[20220406-17:34:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1246)] [MCS Connection Sequence] send channel join confirm
[20220406-17:34:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1239)] [MCS Connection Sequence] receive channel join request
[20220406-17:34:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1246)] [MCS Connection Sequence] send channel join confirm
[20220406-17:34:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1239)] [MCS Connection Sequence] receive channel join request
[20220406-17:34:33] [ERROR] [xrdp_mcs_recv_cjrq(xrdp_mcs.c:746)] Parsed [ITU-T T.125] DomainMCSPDU choice index expected 14, received 25
[20220406-17:34:33] [ERROR] [xrdp_mcs_incoming(xrdp_mcs.c:1242)] [MCS Connection Sequence] receive channel join request failed
[20220406-17:34:33] [ERROR] [xrdp_sec_incoming(xrdp_sec.c:2773)] xrdp_sec_incoming: xrdp_mcs_incoming failed
[20220406-17:34:33] [ERROR] [xrdp_rdp_incoming(xrdp_rdp.c:921)] xrdp_rdp_incoming: xrdp_sec_incoming failed
[20220406-17:34:33] [ERROR] [xrdp_process_main_loop(xrdp_process.c:302)] xrdp_process_main_loop: libxrdp_process_incoming failed
[20220406-17:34:33] [DEBUG] [g_sck_close(os_calls.c:766)] Closed socket 12 (AF_INET 10.4.9.34:3389)
[20220406-17:34:33] [DEBUG] [close_rdp_socket(xrdp_mcs.c:1387)] xrdp_mcs_disconnect - socket closed
```

Alright, here's what I get from enabling TRACE on that channel

```
[20220408-16:31:56] [INFO ] [xrdp_listen_process_startup_params(xrdp_listen.c:719)] setting send buffer to 262144 bytes
[20220408-16:31:56] [INFO ] [xrdp_listen_process_startup_params(xrdp_listen.c:734)] send buffer set to 425984 bytes
[20220408-16:31:56] [INFO ] [xrdp_listen_process_startup_params(xrdp_listen.c:674)] xrdp_listen_pp done
[20220408-16:31:56] [DEBUG] [g_sck_close(os_calls.c:766)] Closed socket 7 (AF_INET 0.0.0.0:3389)
[20220408-16:31:58] [INFO ] [main(xrdp.c:685)] starting xrdp with pid 61029
[20220408-16:31:58] [INFO ] [xrdp_listen_process_startup_params(xrdp_listen.c:678)] address [0.0.0.0] port [3389] mode 1
[20220408-16:31:58] [INFO ] [xrdp_listen_process_startup_params(xrdp_listen.c:687)] listening to port 3389 on 0.0.0.0
[20220408-16:31:58] [INFO ] [xrdp_listen_process_startup_params(xrdp_listen.c:719)] setting send buffer to 262144 bytes
[20220408-16:31:58] [INFO ] [xrdp_listen_process_startup_params(xrdp_listen.c:734)] send buffer set to 425984 bytes
[20220408-16:31:58] [INFO ] [xrdp_listen_process_startup_params(xrdp_listen.c:674)] xrdp_listen_pp done
[20220408-16:33:46] [INFO ] [g_sck_accept(os_calls.c:1395)] Socket 12: AF_INET connection received from 10.13.37.123 port 3564
[20220408-16:33:46] [DEBUG] [g_sck_close(os_calls.c:766)] Closed socket 12 (AF_INET 10.4.9.34:3389)
[20220408-16:33:46] [DEBUG] [g_sck_close(os_calls.c:766)] Closed socket 11 (AF_INET 0.0.0.0:3389)
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ini_version, value 1
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item fork, value true
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item port, value 3389
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item use_vsock, value false
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item tcp_nodelay, value false
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item tcp_keepalive, value true
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item tcp_send_buffer_bytes, value 262144
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item security_layer, value negotiate
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item crypt_level, value low
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item certificate, value 
[20220408-16:33:46] [INFO ] [xrdp_rdp_read_config(xrdp_rdp.c:225)] Using default X.509 certificate: /etc/xrdp/cert.pem
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item key_file, value 
[20220408-16:33:46] [INFO ] [xrdp_rdp_read_config(xrdp_rdp.c:257)] Using default X.509 key file: /etc/xrdp/key.pem
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ssl_protocols, value TLSv1.2, TLSv1.3
[20220408-16:33:46] [DEBUG] [ssl_get_protocols_from_string(ssl_calls.c:1458)] TLSv1.3 enabled
[20220408-16:33:46] [DEBUG] [ssl_get_protocols_from_string(ssl_calls.c:1470)] TLSv1.2 enabled
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item autorun, value 
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item allow_channels, value true
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item allow_multimon, value true
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item bitmap_cache, value true
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item bitmap_compression, value true
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item bulk_compression, value true
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item max_bpp, value 128
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item xserverbpp, value 128
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item use_compression, value yes
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item new_cursors, value true
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item use_fastpath, value both
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item blue, value 009cb5
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item grey, value dedede
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_top_window_bg_color, value 009cb5
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_width, value 350
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_height, value 430
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_bg_color, value dedede
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_logo_filename, value 
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_logo_x_pos, value 55
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_logo_y_pos, value 50
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_label_x_pos, value 30
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_label_width, value 65
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_input_x_pos, value 110
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_input_width, value 210
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_input_y_pos, value 220
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_btn_ok_x_pos, value 142
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_btn_ok_y_pos, value 370
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_btn_ok_width, value 85
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_btn_ok_height, value 30
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_btn_cancel_x_pos, value 237
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_btn_cancel_y_pos, value 370
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_btn_cancel_width, value 85
[20220408-16:33:46] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_btn_cancel_height, value 30
[20220408-16:33:46] [DEBUG] [xrdp_iso_incoming(xrdp_iso.c:485)] [ITU-T X.224] Connection Sequence: receive connection request
[20220408-16:33:46] [INFO ] [xrdp_iso_negotiate_security(xrdp_iso.c:194)] Security protocol: configured [SSL|RDP], requested [RDP], selected [RDP]
[20220408-16:33:46] [DEBUG] [xrdp_iso_incoming(xrdp_iso.c:587)] [ITU-T X.224] Connection Sequence: send connection confirmation
[20220408-16:33:46] [DEBUG] [xrdp_iso_incoming(xrdp_iso.c:594)] [ITU-T X.224] Connection Sequence: completed
[20220408-16:33:46] [DEBUG] [xrdp_sec_incoming(xrdp_sec.c:2709)] Using RDP security, and reading the server configuration
[20220408-16:33:46] [DEBUG] [xrdp_sec_incoming(xrdp_sec.c:2742)] Server config: pub_mod bytes 256
[20220408-16:33:46] [DEBUG] [xrdp_sec_incoming(xrdp_sec.c:2753)] Server config: pri_exp 256
[20220408-16:33:46] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1188)] [MCS Connection Sequence] receive connection request
[20220408-16:33:46] [INFO ] [xrdp_sec_process_mcs_data_CS_CORE(xrdp_sec.c:1972)] Connected client computer name: WT00806465d8c3
[20220408-16:33:46] [DEBUG] [xrdp_sec_process_mcs_data_CS_SECURITY(xrdp_sec.c:2151)] Client supports 40 bit encryption
[20220408-16:33:46] [DEBUG] [xrdp_sec_process_mcs_data_CS_SECURITY(xrdp_sec.c:2155)] Client supports 128 bit encryption
[20220408-16:33:46] [DEBUG] [xrdp_sec_process_mcs_data_CS_SECURITY(xrdp_sec.c:2159)] Client supports 56 bit encryption
[20220408-16:33:46] [DEBUG] [xrdp_sec_process_mcs_data_CS_SECURITY(xrdp_sec.c:2163)] Client supports fips encryption
[20220408-16:33:46] [DEBUG] [xrdp_sec_process_mcs_data_CS_SECURITY(xrdp_sec.c:2225)] Client and server both support low encryption, using RDP 40-bit encryption.
[20220408-16:33:46] [DEBUG] [xrdp_sec_process_mcs_data(xrdp_sec.c:2460)] Received [MS-RDPBCGR] TS_UD_CS_CLUSTER - no-op
[20220408-16:33:46] [TRACE] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2283)] Received [MS-RDPBCGR] TS_UD_CS_NET channelCount 6
[20220408-16:33:46] [TRACE] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2304)] Received [MS-RDPBCGR] TS_UD_CS_NET.CHANNEL_DEF 0, name rdpdr, options 0x82000000
[20220408-16:33:46] [DEBUG] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2310)] Adding channel: name rdpdr, channel id 1004, flags 0x82000000
[20220408-16:33:46] [TRACE] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2304)] Received [MS-RDPBCGR] TS_UD_CS_NET.CHANNEL_DEF 1, name cliprdr, options 0x98000000
[20220408-16:33:46] [DEBUG] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2310)] Adding channel: name cliprdr, channel id 1005, flags 0x98000000
[20220408-16:33:46] [TRACE] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2304)] Received [MS-RDPBCGR] TS_UD_CS_NET.CHANNEL_DEF 2, name CUVA, options 0x88000000
[20220408-16:33:46] [DEBUG] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2310)] Adding channel: name CUVA, channel id 1006, flags 0x88000000
[20220408-16:33:46] [TRACE] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2304)] Received [MS-RDPBCGR] TS_UD_CS_NET.CHANNEL_DEF 3, name AURTCX, options 0x88000000
[20220408-16:33:46] [DEBUG] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2310)] Adding channel: name AURTCX, channel id 1007, flags 0x88000000
[20220408-16:33:46] [TRACE] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2304)] Received [MS-RDPBCGR] TS_UD_CS_NET.CHANNEL_DEF 4, name USBVDX, options 0x88000000
[20220408-16:33:46] [DEBUG] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2310)] Adding channel: name USBVDX, channel id 1008, flags 0x88000000
[20220408-16:33:46] [TRACE] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2304)] Received [MS-RDPBCGR] TS_UD_CS_NET.CHANNEL_DEF 5, name CARANOC, options 0x82000000
[20220408-16:33:46] [DEBUG] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2310)] Adding channel: name CARANOC, channel id 1009, flags 0x82000000
[20220408-16:33:46] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1202)] [MCS Connection Sequence] construct connection reponse
[20220408-16:33:46] [DEBUG] [xrdp_mcs_out_gcc_data(xrdp_mcs.c:1038)] using 2048 bit RSA key
[20220408-16:33:46] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1209)] [MCS Connection Sequence] send connection reponse
[20220408-16:33:46] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1216)] [MCS Connection Sequence] receive erect domain request
[20220408-16:33:46] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1223)] [MCS Connection Sequence] receive attach user request
[20220408-16:33:46] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1230)] [MCS Connection Sequence] send attach user confirm
[20220408-16:33:46] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1239)] [MCS Connection Sequence] receive channel join request
[20220408-16:33:46] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1246)] [MCS Connection Sequence] send channel join confirm
[20220408-16:33:46] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1239)] [MCS Connection Sequence] receive channel join request
[20220408-16:33:46] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1246)] [MCS Connection Sequence] send channel join confirm
[20220408-16:33:46] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1239)] [MCS Connection Sequence] receive channel join request
[20220408-16:33:46] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1246)] [MCS Connection Sequence] send channel join confirm
[20220408-16:33:46] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1239)] [MCS Connection Sequence] receive channel join request
[20220408-16:33:46] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1246)] [MCS Connection Sequence] send channel join confirm
[20220408-16:33:46] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1239)] [MCS Connection Sequence] receive channel join request
[20220408-16:33:47] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1246)] [MCS Connection Sequence] send channel join confirm
[20220408-16:33:47] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1239)] [MCS Connection Sequence] receive channel join request
[20220408-16:33:47] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1246)] [MCS Connection Sequence] send channel join confirm
[20220408-16:33:47] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1239)] [MCS Connection Sequence] receive channel join request
[20220408-16:33:47] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1246)] [MCS Connection Sequence] send channel join confirm
[20220408-16:33:47] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1239)] [MCS Connection Sequence] receive channel join request
[20220408-16:33:47] [ERROR] [xrdp_mcs_recv_cjrq(xrdp_mcs.c:746)] Parsed [ITU-T T.125] DomainMCSPDU choice index expected 14, received 25
[20220408-16:33:47] [ERROR] [xrdp_mcs_incoming(xrdp_mcs.c:1242)] [MCS Connection Sequence] receive channel join request failed
[20220408-16:33:47] [ERROR] [xrdp_sec_incoming(xrdp_sec.c:2773)] xrdp_sec_incoming: xrdp_mcs_incoming failed
[20220408-16:33:47] [ERROR] [xrdp_rdp_incoming(xrdp_rdp.c:921)] xrdp_rdp_incoming: xrdp_sec_incoming failed
[20220408-16:33:47] [ERROR] [xrdp_process_main_loop(xrdp_process.c:302)] xrdp_process_main_loop: libxrdp_process_incoming failed
[20220408-16:33:47] [DEBUG] [g_sck_close(os_calls.c:766)] Closed socket 12 (AF_INET 10.4.9.34:3389)
[20220408-16:33:47] [DEBUG] [close_rdp_socket(xrdp_mcs.c:1387)] xrdp_mcs_disconnect - socket closed


```

Finally got around to applying the patch, there's the latest debug log

Definitely feel like we're getting somewhere, appreciate the persistance.

```
[20220417-09:02:38] [INFO ] [xrdp_listen_process_startup_params(xrdp_listen.c:719)] setting send buffer to 262144 bytes
[20220417-09:02:38] [INFO ] [xrdp_listen_process_startup_params(xrdp_listen.c:734)] send buffer set to 425984 bytes
[20220417-09:02:38] [INFO ] [xrdp_listen_process_startup_params(xrdp_listen.c:674)] xrdp_listen_pp done
[20220417-09:02:38] [DEBUG] [g_sck_close(os_calls.c:766)] Closed socket 7 (AF_INET 0.0.0.0:3389)
[20220417-09:02:40] [INFO ] [main(xrdp.c:685)] starting xrdp with pid 103615
[20220417-09:02:40] [INFO ] [xrdp_listen_process_startup_params(xrdp_listen.c:678)] address [0.0.0.0] port [3389] mode 1
[20220417-09:02:40] [INFO ] [xrdp_listen_process_startup_params(xrdp_listen.c:687)] listening to port 3389 on 0.0.0.0
[20220417-09:02:40] [INFO ] [xrdp_listen_process_startup_params(xrdp_listen.c:719)] setting send buffer to 262144 bytes
[20220417-09:02:40] [INFO ] [xrdp_listen_process_startup_params(xrdp_listen.c:734)] send buffer set to 425984 bytes
[20220417-09:02:40] [INFO ] [xrdp_listen_process_startup_params(xrdp_listen.c:674)] xrdp_listen_pp done
[20220417-09:05:21] [INFO ] [g_sck_accept(os_calls.c:1395)] Socket 12: AF_INET connection received from 10.13.37.123 port 3585
[20220417-09:05:21] [DEBUG] [g_sck_close(os_calls.c:766)] Closed socket 12 (AF_INET 10.4.9.34:3389)
[20220417-09:05:21] [DEBUG] [g_sck_close(os_calls.c:766)] Closed socket 11 (AF_INET 0.0.0.0:3389)
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ini_version, value 1
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item fork, value true
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item port, value 3389
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item use_vsock, value false
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item tcp_nodelay, value false
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item tcp_keepalive, value true
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item tcp_send_buffer_bytes, value 262144
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item security_layer, value negotiate
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item crypt_level, value low
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item certificate, value 
[20220417-09:05:21] [INFO ] [xrdp_rdp_read_config(xrdp_rdp.c:225)] Using default X.509 certificate: /etc/xrdp/cert.pem
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item key_file, value 
[20220417-09:05:21] [INFO ] [xrdp_rdp_read_config(xrdp_rdp.c:257)] Using default X.509 key file: /etc/xrdp/key.pem
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ssl_protocols, value TLSv1.2, TLSv1.3
[20220417-09:05:21] [DEBUG] [ssl_get_protocols_from_string(ssl_calls.c:1458)] TLSv1.3 enabled
[20220417-09:05:21] [DEBUG] [ssl_get_protocols_from_string(ssl_calls.c:1470)] TLSv1.2 enabled
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item autorun, value 
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item allow_channels, value true
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item allow_multimon, value true
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item bitmap_cache, value true
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item bitmap_compression, value true
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item bulk_compression, value true
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item max_bpp, value 128
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item xserverbpp, value 128
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item use_compression, value yes
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item new_cursors, value true
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item use_fastpath, value both
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item blue, value 009cb5
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item grey, value dedede
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_top_window_bg_color, value 009cb5
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_width, value 350
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_height, value 430
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_bg_color, value dedede
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_logo_filename, value 
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_logo_x_pos, value 55
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_logo_y_pos, value 50
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_label_x_pos, value 30
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_label_width, value 65
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_input_x_pos, value 110
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_input_width, value 210
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_input_y_pos, value 220
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_btn_ok_x_pos, value 142
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_btn_ok_y_pos, value 370
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_btn_ok_width, value 85
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_btn_ok_height, value 30
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_btn_cancel_x_pos, value 237
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_btn_cancel_y_pos, value 370
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_btn_cancel_width, value 85
[20220417-09:05:21] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_btn_cancel_height, value 30
[20220417-09:05:21] [DEBUG] [xrdp_iso_incoming(xrdp_iso.c:485)] [ITU-T X.224] Connection Sequence: receive connection request
[20220417-09:05:21] [INFO ] [xrdp_iso_negotiate_security(xrdp_iso.c:194)] Security protocol: configured [SSL|RDP], requested [RDP], selected [RDP]
[20220417-09:05:21] [DEBUG] [xrdp_iso_incoming(xrdp_iso.c:587)] [ITU-T X.224] Connection Sequence: send connection confirmation
[20220417-09:05:21] [DEBUG] [xrdp_iso_incoming(xrdp_iso.c:594)] [ITU-T X.224] Connection Sequence: completed
[20220417-09:05:21] [DEBUG] [xrdp_sec_incoming(xrdp_sec.c:2709)] Using RDP security, and reading the server configuration
[20220417-09:05:21] [DEBUG] [xrdp_sec_incoming(xrdp_sec.c:2742)] Server config: pub_mod bytes 256
[20220417-09:05:21] [DEBUG] [xrdp_sec_incoming(xrdp_sec.c:2753)] Server config: pri_exp 256
[20220417-09:05:21] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1198)] [MCS Connection Sequence] receive connection request
[20220417-09:05:21] [INFO ] [xrdp_sec_process_mcs_data_CS_CORE(xrdp_sec.c:1972)] Connected client computer name: WT00806465d8c3
[20220417-09:05:21] [DEBUG] [xrdp_sec_process_mcs_data_CS_SECURITY(xrdp_sec.c:2151)] Client supports 40 bit encryption
[20220417-09:05:21] [DEBUG] [xrdp_sec_process_mcs_data_CS_SECURITY(xrdp_sec.c:2155)] Client supports 128 bit encryption
[20220417-09:05:21] [DEBUG] [xrdp_sec_process_mcs_data_CS_SECURITY(xrdp_sec.c:2159)] Client supports 56 bit encryption
[20220417-09:05:21] [DEBUG] [xrdp_sec_process_mcs_data_CS_SECURITY(xrdp_sec.c:2163)] Client supports fips encryption
[20220417-09:05:21] [DEBUG] [xrdp_sec_process_mcs_data_CS_SECURITY(xrdp_sec.c:2225)] Client and server both support low encryption, using RDP 40-bit encryption.
[20220417-09:05:21] [DEBUG] [xrdp_sec_process_mcs_data(xrdp_sec.c:2460)] Received [MS-RDPBCGR] TS_UD_CS_CLUSTER - no-op
[20220417-09:05:21] [TRACE] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2283)] Received [MS-RDPBCGR] TS_UD_CS_NET channelCount 6
[20220417-09:05:21] [TRACE] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2304)] Received [MS-RDPBCGR] TS_UD_CS_NET.CHANNEL_DEF 0, name rdpdr, options 0x82000000
[20220417-09:05:21] [DEBUG] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2310)] Adding channel: name rdpdr, channel id 1004, flags 0x82000000
[20220417-09:05:21] [TRACE] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2304)] Received [MS-RDPBCGR] TS_UD_CS_NET.CHANNEL_DEF 1, name cliprdr, options 0x98000000
[20220417-09:05:21] [DEBUG] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2310)] Adding channel: name cliprdr, channel id 1005, flags 0x98000000
[20220417-09:05:21] [TRACE] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2304)] Received [MS-RDPBCGR] TS_UD_CS_NET.CHANNEL_DEF 2, name CUVA, options 0x88000000
[20220417-09:05:21] [DEBUG] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2310)] Adding channel: name CUVA, channel id 1006, flags 0x88000000
[20220417-09:05:21] [TRACE] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2304)] Received [MS-RDPBCGR] TS_UD_CS_NET.CHANNEL_DEF 3, name AURTCX, options 0x88000000
[20220417-09:05:21] [DEBUG] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2310)] Adding channel: name AURTCX, channel id 1007, flags 0x88000000
[20220417-09:05:21] [TRACE] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2304)] Received [MS-RDPBCGR] TS_UD_CS_NET.CHANNEL_DEF 4, name USBVDX, options 0x88000000
[20220417-09:05:21] [DEBUG] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2310)] Adding channel: name USBVDX, channel id 1008, flags 0x88000000
[20220417-09:05:21] [TRACE] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2304)] Received [MS-RDPBCGR] TS_UD_CS_NET.CHANNEL_DEF 5, name CARANOC, options 0x82000000
[20220417-09:05:21] [DEBUG] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2310)] Adding channel: name CARANOC, channel id 1009, flags 0x82000000
[20220417-09:05:21] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1212)] [MCS Connection Sequence] construct connection reponse
[20220417-09:05:21] [DEBUG] [xrdp_mcs_out_gcc_data(xrdp_mcs.c:1048)] using 2048 bit RSA key
[20220417-09:05:21] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1219)] [MCS Connection Sequence] send connection reponse
[20220417-09:05:21] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1226)] [MCS Connection Sequence] receive erect domain request
[20220417-09:05:21] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1233)] [MCS Connection Sequence] receive attach user request
[20220417-09:05:21] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1240)] [MCS Connection Sequence] send attach user confirm
[20220417-09:05:21] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1249)] [MCS Connection Sequence] receive channel join request
[20220417-09:05:21] [TRACE] [xrdp_mcs_recv_cjrq(xrdp_mcs.c:745)] Received [ITU-T T.125] DomainMCSPDU choice index 14 (ChannelJoinRequest)
[20220417-09:05:21] [TRACE] [xrdp_mcs_recv_cjrq(xrdp_mcs.c:783)] Received [ITU-T T.125] ChannelJoinRequest initiator 1, channelId 1003, nonStandard (not present)
[20220417-09:05:21] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1256)] [MCS Connection Sequence] send channel join confirm
[20220417-09:05:21] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1249)] [MCS Connection Sequence] receive channel join request
[20220417-09:05:21] [TRACE] [xrdp_mcs_recv_cjrq(xrdp_mcs.c:745)] Received [ITU-T T.125] DomainMCSPDU choice index 14 (ChannelJoinRequest)
[20220417-09:05:21] [TRACE] [xrdp_mcs_recv_cjrq(xrdp_mcs.c:783)] Received [ITU-T T.125] ChannelJoinRequest initiator 1, channelId 1004, nonStandard (not present)
[20220417-09:05:21] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1256)] [MCS Connection Sequence] send channel join confirm
[20220417-09:05:21] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1249)] [MCS Connection Sequence] receive channel join request
[20220417-09:05:21] [TRACE] [xrdp_mcs_recv_cjrq(xrdp_mcs.c:745)] Received [ITU-T T.125] DomainMCSPDU choice index 14 (ChannelJoinRequest)
[20220417-09:05:21] [TRACE] [xrdp_mcs_recv_cjrq(xrdp_mcs.c:783)] Received [ITU-T T.125] ChannelJoinRequest initiator 1, channelId 1005, nonStandard (not present)
[20220417-09:05:21] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1256)] [MCS Connection Sequence] send channel join confirm
[20220417-09:05:21] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1249)] [MCS Connection Sequence] receive channel join request
[20220417-09:05:21] [TRACE] [xrdp_mcs_recv_cjrq(xrdp_mcs.c:745)] Received [ITU-T T.125] DomainMCSPDU choice index 14 (ChannelJoinRequest)
[20220417-09:05:21] [TRACE] [xrdp_mcs_recv_cjrq(xrdp_mcs.c:783)] Received [ITU-T T.125] ChannelJoinRequest initiator 1, channelId 1006, nonStandard (not present)
[20220417-09:05:21] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1256)] [MCS Connection Sequence] send channel join confirm
[20220417-09:05:21] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1249)] [MCS Connection Sequence] receive channel join request
[20220417-09:05:21] [TRACE] [xrdp_mcs_recv_cjrq(xrdp_mcs.c:745)] Received [ITU-T T.125] DomainMCSPDU choice index 14 (ChannelJoinRequest)
[20220417-09:05:21] [TRACE] [xrdp_mcs_recv_cjrq(xrdp_mcs.c:783)] Received [ITU-T T.125] ChannelJoinRequest initiator 1, channelId 1007, nonStandard (not present)
[20220417-09:05:21] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1256)] [MCS Connection Sequence] send channel join confirm
[20220417-09:05:21] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1249)] [MCS Connection Sequence] receive channel join request
[20220417-09:05:21] [TRACE] [xrdp_mcs_recv_cjrq(xrdp_mcs.c:745)] Received [ITU-T T.125] DomainMCSPDU choice index 14 (ChannelJoinRequest)
[20220417-09:05:21] [TRACE] [xrdp_mcs_recv_cjrq(xrdp_mcs.c:783)] Received [ITU-T T.125] ChannelJoinRequest initiator 1, channelId 1008, nonStandard (not present)
[20220417-09:05:21] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1256)] [MCS Connection Sequence] send channel join confirm
[20220417-09:05:21] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1249)] [MCS Connection Sequence] receive channel join request
[20220417-09:05:21] [TRACE] [xrdp_mcs_recv_cjrq(xrdp_mcs.c:745)] Received [ITU-T T.125] DomainMCSPDU choice index 14 (ChannelJoinRequest)
[20220417-09:05:21] [TRACE] [xrdp_mcs_recv_cjrq(xrdp_mcs.c:783)] Received [ITU-T T.125] ChannelJoinRequest initiator 1, channelId 1009, nonStandard (not present)
[20220417-09:05:21] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1256)] [MCS Connection Sequence] send channel join confirm
[20220417-09:05:21] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1249)] [MCS Connection Sequence] receive channel join request
[20220417-09:05:21] [TRACE] [xrdp_mcs_recv_cjrq(xrdp_mcs.c:745)] Received [ITU-T T.125] DomainMCSPDU choice index 25 (ChannelJoinRequest)
[20220417-09:05:21] [ERROR] [xrdp_mcs_recv_cjrq(xrdp_mcs.c:750)] Parsed [ITU-T T.125] DomainMCSPDU choice index expected 14, received 25
[20220417-09:05:21] [ERROR] [xrdp_mcs_incoming(xrdp_mcs.c:1252)] [MCS Connection Sequence] receive channel join request failed
[20220417-09:05:21] [ERROR] [xrdp_sec_incoming(xrdp_sec.c:2773)] xrdp_sec_incoming: xrdp_mcs_incoming failed
[20220417-09:05:21] [ERROR] [xrdp_rdp_incoming(xrdp_rdp.c:921)] xrdp_rdp_incoming: xrdp_sec_incoming failed
[20220417-09:05:21] [ERROR] [xrdp_process_main_loop(xrdp_process.c:302)] xrdp_process_main_loop: libxrdp_process_incoming failed
[20220417-09:05:21] [DEBUG] [g_sck_close(os_calls.c:766)] Closed socket 12 (AF_INET 10.4.9.34:3389)
[20220417-09:05:21] [DEBUG] [close_rdp_socket(xrdp_mcs.c:1397)] xrdp_mcs_disconnect - socket closed
```

Thanks for the reply.

I checked out the `wyse_thin_client` branch and re-compiled but still unable to connect from the client.  Below is the debug log I gathered

```
[20220510-21:19:58] [INFO ] [xrdp_listen_process_startup_params(xrdp_listen.c:719)] setting send buffer to 262144 bytes
[20220510-21:19:58] [INFO ] [xrdp_listen_process_startup_params(xrdp_listen.c:734)] send buffer set to 425984 bytes
[20220510-21:19:58] [INFO ] [xrdp_listen_process_startup_params(xrdp_listen.c:674)] xrdp_listen_pp done
[20220510-21:19:58] [DEBUG] [g_sck_close(os_calls.c:755)] Closed socket 7 (AF_INET 0.0.0.0:3389)
[20220510-21:20:00] [INFO ] [main(xrdp.c:691)] starting xrdp with pid 197801
[20220510-21:20:00] [INFO ] [xrdp_listen_process_startup_params(xrdp_listen.c:678)] address [0.0.0.0] port [3389] mode 1
[20220510-21:20:00] [INFO ] [xrdp_listen_process_startup_params(xrdp_listen.c:687)] listening to port 3389 on 0.0.0.0
[20220510-21:20:00] [INFO ] [xrdp_listen_process_startup_params(xrdp_listen.c:719)] setting send buffer to 262144 bytes
[20220510-21:20:00] [INFO ] [xrdp_listen_process_startup_params(xrdp_listen.c:734)] send buffer set to 425984 bytes
[20220510-21:20:00] [INFO ] [xrdp_listen_process_startup_params(xrdp_listen.c:674)] xrdp_listen_pp done
[20220510-21:22:33] [INFO ] [g_sck_accept(os_calls.c:1384)] Socket 12: AF_INET connection received from 10.13.37.123 port 3584
[20220510-21:22:33] [DEBUG] [g_sck_close(os_calls.c:755)] Closed socket 12 (AF_INET 10.4.9.34:3389)
[20220510-21:22:33] [DEBUG] [g_sck_close(os_calls.c:755)] Closed socket 11 (AF_INET 0.0.0.0:3389)
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ini_version, value 1
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item fork, value true
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item port, value 3389
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item use_vsock, value false
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item tcp_nodelay, value false
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item tcp_keepalive, value true
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item tcp_send_buffer_bytes, value 262144
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item security_layer, value negotiate
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item crypt_level, value low
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item certificate, value 
[20220510-21:22:33] [INFO ] [xrdp_rdp_read_config(xrdp_rdp.c:225)] Using default X.509 certificate: /etc/xrdp/cert.pem
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item key_file, value 
[20220510-21:22:33] [INFO ] [xrdp_rdp_read_config(xrdp_rdp.c:257)] Using default X.509 key file: /etc/xrdp/key.pem
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ssl_protocols, value TLSv1.2, TLSv1.3
[20220510-21:22:33] [DEBUG] [ssl_get_protocols_from_string(ssl_calls.c:1458)] TLSv1.3 enabled
[20220510-21:22:33] [DEBUG] [ssl_get_protocols_from_string(ssl_calls.c:1470)] TLSv1.2 enabled
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item autorun, value 
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item allow_channels, value true
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item allow_multimon, value true
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item bitmap_cache, value true
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item bitmap_compression, value true
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item bulk_compression, value true
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item max_bpp, value 128
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item xserverbpp, value 128
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item use_compression, value yes
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item new_cursors, value true
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item use_fastpath, value both
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item blue, value 009cb5
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item grey, value dedede
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_top_window_bg_color, value 009cb5
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_width, value 350
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_height, value 430
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_bg_color, value dedede
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_logo_filename, value 
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_logo_x_pos, value 55
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_logo_y_pos, value 50
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_label_x_pos, value 30
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_label_width, value 65
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_input_x_pos, value 110
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_input_width, value 210
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_input_y_pos, value 220
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_btn_ok_x_pos, value 142
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_btn_ok_y_pos, value 370
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_btn_ok_width, value 85
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_btn_ok_height, value 30
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_btn_cancel_x_pos, value 237
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_btn_cancel_y_pos, value 370
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_btn_cancel_width, value 85
[20220510-21:22:33] [DEBUG] [xrdp_rdp_read_config(xrdp_rdp.c:69)] item ls_btn_cancel_height, value 30
[20220510-21:22:33] [DEBUG] [xrdp_iso_incoming(xrdp_iso.c:485)] [ITU-T X.224] Connection Sequence: receive connection request
[20220510-21:22:33] [INFO ] [xrdp_iso_negotiate_security(xrdp_iso.c:194)] Security protocol: configured [SSL|RDP], requested [RDP], selected [RDP]
[20220510-21:22:33] [DEBUG] [xrdp_iso_incoming(xrdp_iso.c:587)] [ITU-T X.224] Connection Sequence: send connection confirmation
[20220510-21:22:33] [DEBUG] [xrdp_iso_incoming(xrdp_iso.c:594)] [ITU-T X.224] Connection Sequence: completed
[20220510-21:22:33] [DEBUG] [xrdp_sec_incoming(xrdp_sec.c:2808)] Using RDP security, and reading the server configuration
[20220510-21:22:33] [DEBUG] [xrdp_sec_incoming(xrdp_sec.c:2841)] Server config: pub_mod bytes 256
[20220510-21:22:33] [DEBUG] [xrdp_sec_incoming(xrdp_sec.c:2852)] Server config: pri_exp 256
[20220510-21:22:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1190)] [MCS Connection Sequence] receive connection request
[20220510-21:22:33] [INFO ] [xrdp_sec_process_mcs_data_CS_CORE(xrdp_sec.c:1974)] Connected client computer name: WT00806465d8c3
[20220510-21:22:33] [DEBUG] [xrdp_sec_process_mcs_data_CS_SECURITY(xrdp_sec.c:2153)] Client supports 40 bit encryption
[20220510-21:22:33] [DEBUG] [xrdp_sec_process_mcs_data_CS_SECURITY(xrdp_sec.c:2157)] Client supports 128 bit encryption
[20220510-21:22:33] [DEBUG] [xrdp_sec_process_mcs_data_CS_SECURITY(xrdp_sec.c:2161)] Client supports 56 bit encryption
[20220510-21:22:33] [DEBUG] [xrdp_sec_process_mcs_data_CS_SECURITY(xrdp_sec.c:2165)] Client supports fips encryption
[20220510-21:22:33] [DEBUG] [xrdp_sec_process_mcs_data_CS_SECURITY(xrdp_sec.c:2227)] Client and server both support low encryption, using RDP 40-bit encryption.
[20220510-21:22:33] [DEBUG] [xrdp_sec_process_mcs_data(xrdp_sec.c:2559)] Received [MS-RDPBCGR] TS_UD_CS_CLUSTER - no-op
[20220510-21:22:33] [TRACE] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2285)] Received [MS-RDPBCGR] TS_UD_CS_NET channelCount 6
[20220510-21:22:33] [TRACE] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2306)] Received [MS-RDPBCGR] TS_UD_CS_NET.CHANNEL_DEF 0, name rdpdr, options 0x82000000
[20220510-21:22:33] [DEBUG] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2312)] Adding channel: name rdpdr, channel id 1004, flags 0x82000000
[20220510-21:22:33] [TRACE] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2306)] Received [MS-RDPBCGR] TS_UD_CS_NET.CHANNEL_DEF 1, name cliprdr, options 0x98000000
[20220510-21:22:33] [DEBUG] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2312)] Adding channel: name cliprdr, channel id 1005, flags 0x98000000
[20220510-21:22:33] [TRACE] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2306)] Received [MS-RDPBCGR] TS_UD_CS_NET.CHANNEL_DEF 2, name CUVA, options 0x88000000
[20220510-21:22:33] [DEBUG] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2312)] Adding channel: name CUVA, channel id 1006, flags 0x88000000
[20220510-21:22:33] [TRACE] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2306)] Received [MS-RDPBCGR] TS_UD_CS_NET.CHANNEL_DEF 3, name AURTCX, options 0x88000000
[20220510-21:22:33] [DEBUG] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2312)] Adding channel: name AURTCX, channel id 1007, flags 0x88000000
[20220510-21:22:33] [TRACE] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2306)] Received [MS-RDPBCGR] TS_UD_CS_NET.CHANNEL_DEF 4, name USBVDX, options 0x88000000
[20220510-21:22:33] [DEBUG] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2312)] Adding channel: name USBVDX, channel id 1008, flags 0x88000000
[20220510-21:22:33] [TRACE] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2306)] Received [MS-RDPBCGR] TS_UD_CS_NET.CHANNEL_DEF 5, name CARANOC, options 0x82000000
[20220510-21:22:33] [DEBUG] [xrdp_sec_process_mcs_data_channels(xrdp_sec.c:2312)] Adding channel: name CARANOC, channel id 1009, flags 0x82000000
[20220510-21:22:33] [DEBUG] [xrdp_mcs_set_user_channel_id(xrdp_mcs.c:1456)] MCS user is 9, channel id is 1010
[20220510-21:22:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1204)] [MCS Connection Sequence] construct connection reponse
[20220510-21:22:33] [DEBUG] [xrdp_mcs_out_gcc_data(xrdp_mcs.c:1040)] using 2048 bit RSA key
[20220510-21:22:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1211)] [MCS Connection Sequence] send connection reponse
[20220510-21:22:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1218)] [MCS Connection Sequence] receive erect domain request
[20220510-21:22:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1225)] [MCS Connection Sequence] receive attach user request
[20220510-21:22:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1232)] [MCS Connection Sequence] send attach user confirm
[20220510-21:22:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1246)] [MCS Connection Sequence] receive channel join request
[20220510-21:22:33] [TRACE] [xrdp_mcs_recv_cjrq(xrdp_mcs.c:742)] Received [ITU-T T.125] DomainMCSPDU choice index 14 (ChannelJoinRequest)
[20220510-21:22:33] [TRACE] [xrdp_mcs_recv_cjrq(xrdp_mcs.c:775)] Received [ITU-T T.125] ChannelJoinRequest initiator 9, channelId 1003, nonStandard (not present)
[20220510-21:22:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1253)] [MCS Connection Sequence] send channel join confirm
[20220510-21:22:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1246)] [MCS Connection Sequence] receive channel join request
[20220510-21:22:33] [TRACE] [xrdp_mcs_recv_cjrq(xrdp_mcs.c:742)] Received [ITU-T T.125] DomainMCSPDU choice index 14 (ChannelJoinRequest)
[20220510-21:22:33] [TRACE] [xrdp_mcs_recv_cjrq(xrdp_mcs.c:775)] Received [ITU-T T.125] ChannelJoinRequest initiator 9, channelId 1004, nonStandard (not present)
[20220510-21:22:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1253)] [MCS Connection Sequence] send channel join confirm
[20220510-21:22:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1246)] [MCS Connection Sequence] receive channel join request
[20220510-21:22:33] [TRACE] [xrdp_mcs_recv_cjrq(xrdp_mcs.c:742)] Received [ITU-T T.125] DomainMCSPDU choice index 14 (ChannelJoinRequest)
[20220510-21:22:33] [TRACE] [xrdp_mcs_recv_cjrq(xrdp_mcs.c:775)] Received [ITU-T T.125] ChannelJoinRequest initiator 9, channelId 1005, nonStandard (not present)
[20220510-21:22:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1253)] [MCS Connection Sequence] send channel join confirm
[20220510-21:22:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1246)] [MCS Connection Sequence] receive channel join request
[20220510-21:22:33] [TRACE] [xrdp_mcs_recv_cjrq(xrdp_mcs.c:742)] Received [ITU-T T.125] DomainMCSPDU choice index 14 (ChannelJoinRequest)
[20220510-21:22:33] [TRACE] [xrdp_mcs_recv_cjrq(xrdp_mcs.c:775)] Received [ITU-T T.125] ChannelJoinRequest initiator 9, channelId 1006, nonStandard (not present)
[20220510-21:22:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1253)] [MCS Connection Sequence] send channel join confirm
[20220510-21:22:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1246)] [MCS Connection Sequence] receive channel join request
[20220510-21:22:33] [TRACE] [xrdp_mcs_recv_cjrq(xrdp_mcs.c:742)] Received [ITU-T T.125] DomainMCSPDU choice index 14 (ChannelJoinRequest)
[20220510-21:22:33] [TRACE] [xrdp_mcs_recv_cjrq(xrdp_mcs.c:775)] Received [ITU-T T.125] ChannelJoinRequest initiator 9, channelId 1007, nonStandard (not present)
[20220510-21:22:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1253)] [MCS Connection Sequence] send channel join confirm
[20220510-21:22:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1246)] [MCS Connection Sequence] receive channel join request
[20220510-21:22:33] [TRACE] [xrdp_mcs_recv_cjrq(xrdp_mcs.c:742)] Received [ITU-T T.125] DomainMCSPDU choice index 14 (ChannelJoinRequest)
[20220510-21:22:33] [TRACE] [xrdp_mcs_recv_cjrq(xrdp_mcs.c:775)] Received [ITU-T T.125] ChannelJoinRequest initiator 9, channelId 1008, nonStandard (not present)
[20220510-21:22:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1253)] [MCS Connection Sequence] send channel join confirm
[20220510-21:22:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1246)] [MCS Connection Sequence] receive channel join request
[20220510-21:22:33] [TRACE] [xrdp_mcs_recv_cjrq(xrdp_mcs.c:742)] Received [ITU-T T.125] DomainMCSPDU choice index 14 (ChannelJoinRequest)
[20220510-21:22:33] [TRACE] [xrdp_mcs_recv_cjrq(xrdp_mcs.c:775)] Received [ITU-T T.125] ChannelJoinRequest initiator 9, channelId 1009, nonStandard (not present)
[20220510-21:22:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1253)] [MCS Connection Sequence] send channel join confirm
[20220510-21:22:33] [DEBUG] [xrdp_mcs_incoming(xrdp_mcs.c:1246)] [MCS Connection Sequence] receive channel join request
[20220510-21:22:33] [TRACE] [xrdp_mcs_recv_cjrq(xrdp_mcs.c:742)] Received [ITU-T T.125] DomainMCSPDU choice index 25 (ChannelJoinRequest)
[20220510-21:22:33] [ERROR] [xrdp_mcs_recv_cjrq(xrdp_mcs.c:747)] Parsed [ITU-T T.125] DomainMCSPDU choice index expected 14, received 25
[20220510-21:22:33] [ERROR] [xrdp_mcs_incoming(xrdp_mcs.c:1249)] [MCS Connection Sequence] receive channel join request failed
[20220510-21:22:33] [ERROR] [xrdp_sec_incoming(xrdp_sec.c:2872)] xrdp_sec_incoming: xrdp_mcs_incoming failed
[20220510-21:22:33] [ERROR] [xrdp_rdp_incoming(xrdp_rdp.c:921)] xrdp_rdp_incoming: xrdp_sec_incoming failed
[20220510-21:22:33] [ERROR] [xrdp_process_main_loop(xrdp_process.c:302)] xrdp_process_main_loop: libxrdp_process_incoming failed
[20220510-21:22:33] [DEBUG] [g_sck_close(os_calls.c:755)] Closed socket 12 (AF_INET 10.4.9.34:3389)
[20220510-21:22:33] [DEBUG] [close_rdp_socket(xrdp_mcs.c:1393)] xrdp_mcs_disconnect - socket closed

```

Really appreciate all the time spent on this so far.

--- a/libxrdp/libxrdp.h
+++ b/libxrdp/libxrdp.h
@@ -67,6 +67,9 @@ struct xrdp_mcs
     struct stream *client_mcs_data;
     struct stream *server_mcs_data;
     struct list *channel_list;
+    /* This boolean is set to indicate we're expecing channel join
+     * requests as part of the connect sequence */
+    int expecting_channel_join_requests;
 };
 
 /* fastpath */

--- a/libxrdp/xrdp_mcs.c
+++ b/libxrdp/xrdp_mcs.c
@@ -27,6 +27,11 @@
 #include "ms-rdpbcgr.h"
 #include "log.h"
 
+/* Forward references */
+static int
+handle_non_tls_client_channel_join_requests(struct xrdp_mcs *self,
+        struct stream *s, int *appid);
+
 /*****************************************************************************/
 struct xrdp_mcs *
 xrdp_mcs_create(struct xrdp_sec *owner, struct trans *trans,
@@ -124,6 +129,34 @@ xrdp_mcs_send_cjcf(struct xrdp_mcs *self, int userid, int chanid)
 }
 
 /*****************************************************************************/
+/* Reads the header of a DomainMCSPDU and gets the appid
+ *
+ * @return 0 for success */
+static int
+get_domain_mcs_pdu_header(struct xrdp_mcs *self, struct stream *s, int *appid)
+{
+    int opcode;
+    if (xrdp_iso_recv(self->iso_layer, s) != 0)
+    {
+        LOG(LOG_LEVEL_ERROR, "xrdp_mcs_recv: xrdp_iso_recv failed");
+        return 1;
+    }
+
+    if (!s_check_rem_and_log(s, 1, "Parsing [ITU-T T.125] DomainMCSPDU"))
+    {
+        return 1;
+    }
+
+    /* The DomainMCSPDU choice index is a 6-bit int with the 2 least
+       significant bits of the byte as padding */
+    in_uint8(s, opcode);
+    *appid = opcode >> 2; /* 2-bit padding */
+    LOG_DEVEL(LOG_LEVEL_TRACE,
+              "Received [ITU-T T.125] DomainMCSPDU choice index %d", *appid);
+    return 0;
+}
+
+/*****************************************************************************/
 /*
  * Processes an [ITU-T T.125] DomainMCSPDU message.
  *
@@ -136,69 +169,31 @@ int
 xrdp_mcs_recv(struct xrdp_mcs *self, struct stream *s, int *chan)
 {
     int appid;
-    int opcode;
     int len;
-    int userid;
-    int chanid;
 
-
-    while (1)
+    if (get_domain_mcs_pdu_header(self, s, &appid) != 0)
     {
-        if (xrdp_iso_recv(self->iso_layer, s) != 0)
-        {
-            LOG(LOG_LEVEL_ERROR, "xrdp_mcs_recv: xrdp_iso_recv failed");
-            return 1;
-        }
-
-        if (!s_check_rem_and_log(s, 1, "Parsing [ITU-T T.125] DomainMCSPDU"))
-        {
-            return 1;
-        }
-
-        /* The DomainMCSPDU choice index is a 6-bit int with the 2 least
-           significant bits of the byte as padding */
-        in_uint8(s, opcode);
-        appid = opcode >> 2; /* 2-bit padding */
-        LOG_DEVEL(LOG_LEVEL_TRACE,
-                  "Received [ITU-T T.125] DomainMCSPDU choice index %d", appid);
-
-        if (appid == MCS_DPUM) /* Disconnect Provider Ultimatum */
-        {
-            LOG_DEVEL(LOG_LEVEL_TRACE, "Received [ITU-T T.125] DisconnectProviderUltimatum");
-            LOG(LOG_LEVEL_DEBUG, "Recieved disconnection request");
-            return 1;
-        }
-
-        /* MCS_CJRQ: Channel Join ReQuest
-           this is channels getting added from the client */
-        if (appid == MCS_CJRQ)
-        {
-            if (!s_check_rem_and_log(s, 4, "Parsing [ITU-T T.125] ChannelJoinRequest"))
-            {
-                return 1;
-            }
-
-            in_uint16_be(s, userid);
-            in_uint16_be(s, chanid);
-            LOG_DEVEL(LOG_LEVEL_TRACE, "Received [ITU-T T.125] ChannelJoinRequest "
-                      "initiator 0x%4.4x, channelId 0x%4.4x",  userid, chanid);
-
-            if (xrdp_mcs_send_cjcf(self, userid, chanid) != 0)
-            {
-                LOG(LOG_LEVEL_WARNING, "[ITU-T T.125] Channel join sequence: failed");
-            }
-
-            s = libxrdp_force_read(self->iso_layer->trans);
-            if (s == 0)
-            {
-                LOG(LOG_LEVEL_ERROR, "xrdp_mcs_recv: libxrdp_force_read failed");
-                return 1;
-            }
-
-            continue;
-        }
-        break;
+        return 1;
     }
+
+    if (self->expecting_channel_join_requests)
+    {
+        if (handle_non_tls_client_channel_join_requests(self, s, &appid) != 0)
+        {
+            return 1;
+        }
+        LOG(LOG_LEVEL_DEBUG, "[MCS Connection Sequence] completed");
+        self->expecting_channel_join_requests = 0;
+    }
+
+    if (appid == MCS_DPUM) /* Disconnect Provider Ultimatum */
+    {
+        LOG_DEVEL(LOG_LEVEL_TRACE,
+                  "Received [ITU-T T.125] DisconnectProviderUltimatum");
+        LOG(LOG_LEVEL_DEBUG, "Recieved disconnection request");
+        return 1;
+    }
+
 
     if (appid != MCS_SDRQ)
     {
@@ -1177,6 +1172,127 @@ xrdp_mcs_send_connect_response(struct xrdp_mcs *self)
 }
 
 /*****************************************************************************/
+/* Handle all client channel join requests for a non-TLS connection
+ *
+ * @param self MCS structure
+ * @param s Input stream
+ * @param[in,out] appid Type of the MCS PDU whose header has just been read.
+ * @return 0 for success
+ *
+ * For non-TLS connections, the channel join MCS PDUs are followed by
+ * another MCS PDU. This not the case for TLS connections.
+ *
+ * Called when an MCS PDU header has been read, but the PDU has not
+ * been processed.
+ *
+ * If the PDU is a channel join request, it is processed, and the next
+ * PDU header is read. When we've exhausted all the channel join requests,
+ * the type of the next PDU is passed back to the caller for the caller
+ * to process.
+ *
+ * In order to cater for older clients which may not conform exactly to
+ * the specification, we simply take all the join requests which come in,
+ * and respond to them.
+ *
+ * See :-
+ * - https://github.com/neutrinolabs/xrdp/issues/2166
+ * - [MS-RDPBCGR] 3.2.5.3.8 and 3.2.5.3.8
+ */
+static int
+handle_non_tls_client_channel_join_requests(struct xrdp_mcs *self,
+        struct stream *s, int *appid)
+{
+    int rv = 0;
+    while (*appid == MCS_CJRQ)
+    {
+        int userid;
+        int chanid;
+
+        if (!s_check_rem_and_log(s, 4, "Parsing [ITU-T T.125] "
+                                 "ChannelJoinRequest"))
+        {
+            rv = 1;
+            break;
+        }
+
+        in_uint16_be(s, userid);
+        in_uint16_be(s, chanid);
+        LOG_DEVEL(LOG_LEVEL_TRACE,
+                  "Received [ITU-T T.125] ChannelJoinRequest "
+                  "initiator 0x%4.4x, channelId 0x%4.4x", userid, chanid);
+
+        if (xrdp_mcs_send_cjcf(self, userid, chanid) != 0)
+        {
+            LOG(LOG_LEVEL_WARNING,
+                "[ITU-T T.125] Channel join sequence: failed");
+        }
+
+        /* Get the next PDU header */
+        s = libxrdp_force_read(self->iso_layer->trans);
+        if (s == 0)
+        {
+            LOG(LOG_LEVEL_ERROR, "xrdp_mcs_recv: libxrdp_force_read failed");
+            rv = 1;
+            break;
+        }
+        if (get_domain_mcs_pdu_header(self, s, appid) != 0)
+        {
+            rv = 1;
+            break;
+        }
+    }
+
+    return rv;
+}
+
+/*****************************************************************************/
+/* Handle all client channel join requests for a TLS connection
+ *
+ * @param self MCS structure
+ * @return 0 for success
+ *
+ * When were are about to negotiate a TLS connection, it is important that
+ * we agree on the exact number of client join request / client join confirm
+ * PDUs, so that we get the TLS 'client hello' message exactly when
+ * expected.
+ *
+ * See [MS-RDPBCGR] 3.2.5.3.8 and 3.2.5.3.8
+ */
+static int
+handle_tls_client_channel_join_requests(struct xrdp_mcs *self)
+{
+    int index;
+    int rv = 0;
+
+    static const char *tag = "[MCS Connection Sequence (TLS)]";
+    /*
+     * Expect a channel join request PDU for each of the static virtual
+     * channels, plus the user channel (self->chanid) and the I/O channel
+     * (MCS_GLOBAL_CHANNEL) */
+    for (index = 0; index < self->channel_list->count + 2; index++)
+    {
+        int channel_id;
+        LOG(LOG_LEVEL_DEBUG, "%s receive channel join request", tag);
+        if (xrdp_mcs_recv_cjrq(self, &channel_id) != 0)
+        {
+            LOG(LOG_LEVEL_ERROR, "%s receive channel join request failed", tag);
+            rv = 1;
+            break;
+        }
+
+        LOG(LOG_LEVEL_DEBUG, "%s send channel join confirm", tag);
+        if (xrdp_mcs_send_cjcf(self, self->userid, channel_id) != 0)
+        {
+            LOG(LOG_LEVEL_ERROR, "%s send channel join confirm failed", tag);
+            rv = 1;
+            break;
+        }
+    }
+
+    return rv;
+}
+
+/*****************************************************************************/
 /* Process and send the MCS messages for the RDP Connection Sequence
  * [MS-RDPBCGR] 1.3.1.1
  *
@@ -1185,8 +1301,6 @@ xrdp_mcs_send_connect_response(struct xrdp_mcs *self)
 int
 xrdp_mcs_incoming(struct xrdp_mcs *self)
 {
-    int index;
-
     LOG(LOG_LEVEL_DEBUG, "[MCS Connection Sequence] receive connection request");
     if (xrdp_mcs_recv_connect_initial(self) != 0)
     {
@@ -1236,29 +1350,25 @@ xrdp_mcs_incoming(struct xrdp_mcs *self)
         return 1;
     }
 
-    /*
-     * Expect a channel join request PDU for each of the static virtual channels, plus
-     * the user channel (self->chanid) and the I/O channel (MCS_GLOBAL_CHANNEL)
-     */
-    for (index = 0; index < self->channel_list->count + 2; index++)
+    if (self->iso_layer->selectedProtocol > PROTOCOL_RDP)
     {
-        int channel_id;
-        LOG(LOG_LEVEL_DEBUG, "[MCS Connection Sequence] receive channel join request");
-        if (xrdp_mcs_recv_cjrq(self, &channel_id) != 0)
+        /* TLS connection. Client and server have to agree on MCS channel
+         * join messages, and these have to be processed before the TLS
+         * client hello */
+        if (handle_tls_client_channel_join_requests(self) != 0)
         {
-            LOG(LOG_LEVEL_ERROR, "[MCS Connection Sequence] receive channel join request failed");
             return 1;
         }
 
-        LOG(LOG_LEVEL_DEBUG, "[MCS Connection Sequence] send channel join confirm");
-        if (xrdp_mcs_send_cjcf(self, self->userid, channel_id) != 0)
-        {
-            LOG(LOG_LEVEL_ERROR, "[MCS Connection Sequence] send channel join confirm failed");
-            return 1;
-        }
+        LOG(LOG_LEVEL_DEBUG, "[MCS Connection Sequence (TLS)] completed");
+    }
+    else
+    {
+        /* Non-TLS connection - channel joins handled in MCS PDU
+         * processing loop */
+        self->expecting_channel_join_requests = 1;
     }
 
-    LOG(LOG_LEVEL_DEBUG, "[MCS Connection Sequence] completed");
     return 0;
 }
 

--- a/libxrdp/xrdp_mcs.c
+++ b/libxrdp/xrdp_mcs.c
@@ -718,6 +718,8 @@ xrdp_mcs_recv_cjrq(struct xrdp_mcs *self)
 {
     int opcode;
     struct stream *s;
+    int initiator;
+    int channel_id;
 
     s = libxrdp_force_read(self->iso_layer->trans);
     if (s == 0)
@@ -753,8 +755,8 @@ xrdp_mcs_recv_cjrq(struct xrdp_mcs *self)
         return 1;
     }
 
-    in_uint8s(s, 4); /* initiator (2 bytes)
-                        channelId (2 bytes) */
+    in_uint16_be(s, initiator);
+    in_uint16_be(s, channel_id);
 
     /*
      * [MS-RDPBCGR] 2.2.1.8 says that the mcsAUrq field is 5 bytes (which have
@@ -772,8 +774,9 @@ xrdp_mcs_recv_cjrq(struct xrdp_mcs *self)
     }
 
     LOG_DEVEL(LOG_LEVEL_TRACE, "Received [ITU-T T.125] ChannelJoinRequest "
-              "initiator (ignored), channelId (ignored), "
+              "initiator %d, channelId %d, "
               "nonStandard (%s)",
+              initiator, channel_id,
               (opcode & 2) ? "present" : "not present");
 
     if (!s_check_end_and_log(s, "MCS protocol error [ITU-T T.125] ChannelJoinRequest"))

--- a/libxrdp/xrdp_mcs.c
+++ b/libxrdp/xrdp_mcs.c
@@ -109,7 +109,7 @@ xrdp_mcs_send_cjcf(struct xrdp_mcs *self, int userid, int chanid)
     out_uint16_be(s, chanid); /* channelId (OPTIONAL) */
     s_mark_end(s);
 
-    LOG_DEVEL(LOG_LEVEL_TRACE, "Sening [ITU-T T.125] ChannelJoinConfirm "
+    LOG_DEVEL(LOG_LEVEL_TRACE, "Sending [ITU-T T.125] ChannelJoinConfirm "
               "result SUCCESS, initiator %d, requested %d, "
               "channelId %d",  userid, chanid, chanid);
     if (xrdp_iso_send(self->iso_layer, s) != 0)
@@ -714,12 +714,11 @@ xrdp_mcs_send_aucf(struct xrdp_mcs *self)
  *
  * returns error */
 static int
-xrdp_mcs_recv_cjrq(struct xrdp_mcs *self)
+xrdp_mcs_recv_cjrq(struct xrdp_mcs *self, int *channel_id)
 {
     int opcode;
     struct stream *s;
     int initiator;
-    int channel_id;
 
     s = libxrdp_force_read(self->iso_layer->trans);
     if (s == 0)
@@ -756,7 +755,7 @@ xrdp_mcs_recv_cjrq(struct xrdp_mcs *self)
     }
 
     in_uint16_be(s, initiator);
-    in_uint16_be(s, channel_id);
+    in_uint16_be(s, *channel_id);
 
     /*
      * [MS-RDPBCGR] 2.2.1.8 says that the mcsAUrq field is 5 bytes (which have
@@ -776,7 +775,7 @@ xrdp_mcs_recv_cjrq(struct xrdp_mcs *self)
     LOG_DEVEL(LOG_LEVEL_TRACE, "Received [ITU-T T.125] ChannelJoinRequest "
               "initiator %d, channelId %d, "
               "nonStandard (%s)",
-              initiator, channel_id,
+              initiator, *channel_id,
               (opcode & 2) ? "present" : "not present");
 
     if (!s_check_end_and_log(s, "MCS protocol error [ITU-T T.125] ChannelJoinRequest"))
@@ -1237,18 +1236,22 @@ xrdp_mcs_incoming(struct xrdp_mcs *self)
         return 1;
     }
 
+    /*
+     * Expect a channel join request PDU for each of the static virtual channels, plus
+     * the user channel (self->chanid) and the I/O channel (MCS_GLOBAL_CHANNEL)
+     */
     for (index = 0; index < self->channel_list->count + 2; index++)
     {
+        int channel_id;
         LOG(LOG_LEVEL_DEBUG, "[MCS Connection Sequence] receive channel join request");
-        if (xrdp_mcs_recv_cjrq(self) != 0)
+        if (xrdp_mcs_recv_cjrq(self, &channel_id) != 0)
         {
             LOG(LOG_LEVEL_ERROR, "[MCS Connection Sequence] receive channel join request failed");
             return 1;
         }
 
         LOG(LOG_LEVEL_DEBUG, "[MCS Connection Sequence] send channel join confirm");
-        if (xrdp_mcs_send_cjcf(self, self->userid,
-                               self->userid + MCS_USERCHANNEL_BASE + index) != 0)
+        if (xrdp_mcs_send_cjcf(self, self->userid, channel_id) != 0)
         {
             LOG(LOG_LEVEL_ERROR, "[MCS Connection Sequence] send channel join confirm failed");
             return 1;

--- a/libxrdp/xrdp_sec.c
+++ b/libxrdp/xrdp_sec.c
@@ -2267,6 +2267,7 @@ xrdp_sec_process_mcs_data_channels(struct xrdp_sec *self, struct stream *s)
     int index;
     struct xrdp_client_info *client_info;
     struct mcs_channel_item *channel_item;
+    int next_mcs_channel_id;
 
     client_info = &(self->rdp_layer->client_info);
     /* this is an option set in xrdp.ini */
@@ -2288,6 +2289,13 @@ xrdp_sec_process_mcs_data_channels(struct xrdp_sec *self, struct stream *s)
             "max 31, received %d", num_channels);
         return 1;
     }
+
+    /* GOTCHA: When adding a channel the MCS channel ID is set to
+     * MCS_GLOBAL_CHANNEL + (index + 1). This is assumed by
+     * xrdp_channel_process(), when mapping an incoming PDU into an
+     * entry in this array */
+    next_mcs_channel_id = MCS_GLOBAL_CHANNEL + 1;
+
     for (index = 0; index < num_channels; index++)
     {
         channel_item = g_new0(struct mcs_channel_item, 1);
@@ -2304,7 +2312,7 @@ xrdp_sec_process_mcs_data_channels(struct xrdp_sec *self, struct stream *s)
             LOG_DEVEL(LOG_LEVEL_TRACE, "Received [MS-RDPBCGR] "
                       "TS_UD_CS_NET.CHANNEL_DEF %d, name %s, options 0x%8.8x",
                       index, channel_item->name, channel_item->flags);
-            channel_item->chanid = MCS_GLOBAL_CHANNEL + (index + 1);
+            channel_item->chanid = next_mcs_channel_id++;
             list_add_item(self->mcs_layer->channel_list,
                           (intptr_t) channel_item);
             LOG(LOG_LEVEL_DEBUG,
@@ -2322,6 +2330,13 @@ xrdp_sec_process_mcs_data_channels(struct xrdp_sec *self, struct stream *s)
             g_free(channel_item);
         }
     }
+
+    /* Set the user channel as well */
+    self->mcs_layer->chanid = next_mcs_channel_id++;
+    self->mcs_layer->userid = self->mcs_layer->chanid - MCS_USERCHANNEL_BASE;
+    LOG_DEVEL(LOG_LEVEL_DEBUG, "MCS user is %d, channel id is %d",
+              self->mcs_layer->userid, self->mcs_layer->chanid);
+
     return 0;
 }
 

--- a/libxrdp/xrdp_sec.c
+++ b/libxrdp/xrdp_sec.c
@@ -1944,6 +1944,7 @@ xrdp_sec_send_fastpath(struct xrdp_sec *self, struct stream *s)
 static int
 xrdp_sec_process_mcs_data_CS_CORE(struct xrdp_sec *self, struct stream *s)
 {
+    int version;
     int colorDepth;
     int postBeta2ColorDepth;
     int highColorDepth;
@@ -1951,8 +1952,10 @@ xrdp_sec_process_mcs_data_CS_CORE(struct xrdp_sec *self, struct stream *s)
     int earlyCapabilityFlags;
     char clientName[INFO_CLIENT_NAME_BYTES / 2] = { '\0' };
 
+    UNUSED_VAR(version);
+
     /* TS_UD_CS_CORE requiered fields */
-    in_uint8s(s, 4); /* version */
+    in_uint32_le(s, version);
     in_uint16_le(s, self->rdp_layer->client_info.display_sizes.session_width);
     in_uint16_le(s, self->rdp_layer->client_info.display_sizes.session_height);
     in_uint16_le(s, colorDepth);
@@ -1975,12 +1978,13 @@ xrdp_sec_process_mcs_data_CS_CORE(struct xrdp_sec *self, struct stream *s)
     in_uint8s(s, 4); /* keyboardFunctionKey */
     in_uint8s(s, 64); /* imeFileName */
     LOG_DEVEL(LOG_LEVEL_TRACE, "Received [MS-RDPBCGR] TS_UD_CS_CORE "
-              "<Requiered fields> version (ignored), desktopWidth %d, "
+              "<Required fields> version %08x, desktopWidth %d, "
               "desktopHeight %d, colorDepth %s, SASSequence (ingored), "
               "keyboardLayout (ignored), clientBuild (ignored), "
               "clientName %s, keyboardType (ignored), "
               "keyboardSubType (ignored), keyboardFunctionKey (ignored), "
-              "imeFileName (ignroed)",
+              "imeFileName (ignored)",
+              version,
               self->rdp_layer->client_info.display_sizes.session_width,
               self->rdp_layer->client_info.display_sizes.session_height,
               (colorDepth == 0xca00 ? "RNS_UD_COLOR_4BPP" :

